### ISSUE TITLE
clang-tidy: [cppcoreguidelines-pro-type-union-access] sometimes does not report location

### ISSUE THREAD

**Issue by ksromanov** [llvm/llvm-project#102945]
[cppcoreguidelines-pro-type-union-access] gets triggered on any union member access. It uses `getMemberLoc()` function to retrieve the location of the union member (i.e. `unionMemberA` in `auto i = Z.unionMemberA`) 
here — https://github.com/llvm/llvm-project/blob/11aa31f595325d6b2dede3364e4b86d78fffe635/clang-tools-extra/clang-tidy/cppcoreguidelines/ProTypeUnionAccessCheck.cpp#L26

Sometimes this location is unavailable. For example

```c++
union Z {
  int a;

  template <class TP> operator TP *() const;
};

Z foo();

int main() {
    void *ret = foo(); // Triggers warning
    return 0;
}
```

This results in the confusing message, _note the absence_ of `file:line:col` triple:

```console
warning: do not access members of unions; use (boost::)variant instead [cppcoreguidelines-pro-type-union-access]
```

_____________________

There are two options to fix the issue:

1. Disable report if the member location is invalid:

```c++
 if (auto MemberLoc = Matched->getMemberLoc(); MemberLoc.isValid())
     diag(MemberLoc,
          "do not access members of unions; use (boost::)variant instead");
```

2. Use `Matched` location instead:

```c++
 if (auto MemberLoc = Matched->getMemberLoc(); MemberLoc.isValid())
     diag(MemberLoc,
          "do not access members of unions; use (boost::)variant instead");
 else
     diag(Matched->getExprLoc(),
          "do not access members of unions; use (boost::)variant instead");
```

or similar code.

Please confirm and advise.

**Comment by ccotter** [comment]
In the spirit of the cppcoreguidelines, I think option 2 would make sense, although the message "use variant" makes less sense in such a case as it's not clear whether it can be used as a *direct* replacement.

**Comment by PiotrZSL** [comment]
"use " could be changed into "consider using", and would be fine.
